pkg/reconciler/identity/reconciler.go

```diff
@@ -155,87 +159,35 @@ func (i *Identity) createServiceAccount(ctx context.Context, namespace, gService
 
 // TODO he iam policy binding should be mocked so that we can unit test it. issue https://github.com/google/knative-gcp/issues/657
 // addIamPolicyBinding will add iam policy binding, which is related to a provided k8s ServiceAccount, to a GCP ServiceAccount.
-func addIamPolicyBinding(ctx context.Context, projectID, gServiceAccount string, kServiceAccount *corev1.ServiceAccount) error {
-	if err := setIamPolicy(ctx, Add, projectID, gServiceAccount, kServiceAccount); err != nil {
-		return err
-	}
-	return nil
-}
-
-// removeIamPolicyBinding will remove iam policy binding, which is related to a provided k8s ServiceAccount, from a GCP ServiceAccount.
-func removeIamPolicyBinding(ctx context.Context, projectID, gServiceAccount string, kServiceAccount *corev1.ServiceAccount) error {
-	if err := setIamPolicy(ctx, Remove, projectID, gServiceAccount, kServiceAccount); err != nil {
-		return err
-	}
-	return nil
-}
-
-func setIamPolicy(ctx context.Context, action, projectID string, gServiceAccount string, kServiceAccount *corev1.ServiceAccount) error {
-	iamService, err := iam.NewService(ctx)
-	if err != nil {
-		return fmt.Errorf("failed to set google iam service: %w", err)
-	}
-
+func (i *Identity) addIamPolicyBinding(ctx context.Context, projectID, gServiceAccount string, kServiceAccount *corev1.ServiceAccount) error {
 	projectId, err := utils.ProjectID(projectID)
 	if err != nil {
 		return fmt.Errorf("failed to get project id: %w", err)
 	}
 
 	// Extract gServiceAccount's project name. The format of gServiceAccountName is service-account-name@project-id.iam.gserviceaccount.com.
 	gsaProject := strings.Split(strings.Split(gServiceAccount, "@")[1], ".")[0]
```

Comment:
Can any malformed string get this far in? If so, I would prefer it returns errors on a bad string, rather than crash.

Reply:
I'm not sure, I didn't modify this part of the code. IAM is able to infer project from the service account which seems to be what we're trying to do here, so I will try modifying the policy manager to use a project wildcard.